python/lsst/pipe/tasks/processCcd.py

```diff
@@ -72,15 +129,112 @@ def run(self, sensorRef):
         """
         self.log.info("Processing %s" % (sensorRef.dataId))
 
+        # initialize postIsrExposure exposure
+        postIsrExposure = self.setPostIsrExposure(sensorRef)
+
         # initialize outputs
-        postIsrExposure = None
-        
-        if self.config.doIsr:
-            postIsrExposure = self.isr.runDataRef(sensorRef).exposure
-        elif self.config.doCalibrate:
-            postIsrExposure = sensorRef.get(self.dataPrefix + "postISRCCD")
-        
+        idFactory = None
+        calib = None
+        sources = None
+        backgrounds = afwMath.BackgroundList()
+        if self.config.doCalibrate:
+            idFactory = self.makeIdFactory(sensorRef)
+            calib = self.calibrate.run(postIsrExposure, idFactory=idFactory)
+            calExposure = calib.exposure
+            if self.config.doWriteCalibrate:
+                sensorRef.put(calib.sources, self.dataPrefix + "icSrc")
+                if calib.matches is not None and self.config.doWriteCalibrateMatches:
+                    normalizedMatches = afwTable.packMatches(calib.matches)
+                    normalizedMatches.table.setMetadata(calib.matchMeta)
+                    sensorRef.put(normalizedMatches, self.dataPrefix + "icMatch")
+            try:
+                for bg in calib.backgrounds:
+                    backgrounds.append(bg)
+            except TypeError:
+                backgrounds.append(calib.backgrounds)
+            except AttributeError:
+                self.log.warn("The calibration task did not return any backgrounds. " +
+                    "Any background subtracted in the calibration process cannot be persisted.")
+        elif sensorRef.datasetExists("calexp"):
+            calExposure = sensorRef.get("calexp", immediate=True)
+        else:
+            raise RuntimeError("No calibrated exposure available for processing")
+
         # delegate most of the work to ProcessImageTask
-        result = self.process(sensorRef, postIsrExposure)
-        result.postIsrExposure = postIsrExposure
-        return result
+        result = self.process(sensorRef, calExposure, idFactory=idFactory)
+
+        # combine the differential background we estimated while detecting the main src catalog
+        # with the background estimated in the calibrate step
+        for bg in result.backgrounds:
+            backgrounds.append(bg)
+        result.backgrounds = backgrounds
```

Comment:
This seems convoluted, and I'm not sure if it needs to be. Is the ordering of the backgrounds within the list important? If so, can we add a comment to say so? If not, can we just `result.backgrounds.extend(backgrounds)` and then use `result.backgrounds` in the calls to `write` and `restoreBackgrounds` below?